gear/gear/database.py

```diff
@@ -1,11 +1,59 @@
 import os
 import json
+import pymysql
 import aiomysql
 import logging
 
+from hailtop.utils import sleep_and_backoff
+
+
 log = logging.getLogger('gear.database')
 
 
+# 1213 - Deadlock found when trying to get lock; try restarting transaction
+retry_codes = (1213,)
+
+
+def retry_async_generator(f):
+    async def wrapper(*args, **kwargs):
+        delay = 0.1
+        errors = 0
+        while True:
+            try:
+                async for x in f(*args, **kwargs):
+                    yield x
+                return
+            except pymysql.err.OperationalError as e:
+                errors += 1
+                if errors % 10 == 0:
+                    log.warning(f'encountered {errors} errors, most recent one was {e}', exc_info=True)
+                if e.args[0] in retry_codes:
+                    pass
+                else:
+                    raise
+            delay = await sleep_and_backoff(delay)
+    return wrapper
+
+
+def retry_coroutine(f):
+    async def wrapper(*args, **kwargs):
+        delay = 0.1
+        errors = 0
+        while True:
+            try:
+                return await f(*args, **kwargs)
+            except pymysql.err.OperationalError as e:
+                errors += 1
+                if errors % 10 == 0:
+                    log.warning(f'encountered {errors} errors, most recent one was {e}', exc_info=True)
+                if e.args[0] in retry_codes:
+                    pass
+                else:
+                    raise
+            delay = await sleep_and_backoff(delay)
+    return wrapper
```

Comment:
In practice, deadlock errors seem rather rare. I've been using this definition:
```python
async def retry_deadlocks(f):
    delay = 0.1
    while True:
        try:
            return await f()
        except pymysql.err.OperationalError as exc:
            if exc.args[0] == 1213:
                log.warn(f'retrying deadlock: {exc}', exc_info=True)
            else:
                raise
        delay = await sleep_and_backoff(delay)
```
I don't think limiting the number of printed errors is necessary here is necessary here.

Reply:
They are not rare for my branch that is trying to schedule in parallel. There I'd say > 25% of calls are deadlock errors.